src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala

```diff
@@ -440,6 +440,12 @@ class VariantSampleMatrix[T](val metadata: VariantMetadata,
       }
   }
 
+  /**
+    * The function {@code f} must be monotonic with respect to the ordering on {@code Locus}
```

Comment:
I don't think this quite captures the complexity of the constraints on `f`.

Let's make ordered keys a typeclass:

```
trait OrderedKey[K] {
  implicit def kOrd: Ordering[K]
  
  type PartitionKey
  implicit def pkOrd: Ordering[PartitionKey]
  
  // monotonic
  def project(key: K): PartitionKey

    // going to be necessary
    implicit def pkct: ClassTag[PartitionKey]
}
```

Then we can define a map of `OrderedKey`s:

```
trait OrderedKeyFunction2[K1, K2](implicit k1ok: OrderedKey[K1], k2ok: OrderedKey[K2]) {
  // monotonic
  def f(k1: K1): K2
  
  // monotonic and commuting: partitionF k1ok.project = k2ok.project f
  def partitionF(k1ok.PartitionKey): k2ok.PartitionKey
}
```

This is probably what `mapMonotonic` should apply.

The `flatMap` situation is slightly more complicated.  I think we need:

 - the restriction of`f` to keys is monotonic over all outputs, namely, if v1 < v2, f(v1) = (v1a, v1b, ...) and f(v2) = (v2p, v2q, ...), then v1x < v1y for all x = a, b, ... and y = p, q, ...

 - project(v1) = project(v1x) (and similarly for v2)

 - the outputs are sorted, v1a <= v1b <= ... (and similarly for v2)